File: api/schemas/tasks.py
from typing import Literal, Optional

from pydantic import BaseModel, Field


TaskType = Literal["view_post"]
TaskStatus = Literal["available", "in_progress", "completed", "blocked"]


class TaskListItem(BaseModel):
    id: int
    type: TaskType = "view_post"
    title: str
    description: Optional[str] = None
    reward: float
    status: TaskStatus = "available"

    chat_id: Optional[str] = None
    channel_post_id: Optional[int] = None
    post_url: Optional[str] = None

    already_completed: bool = False
    can_claim: bool = False
    hold_seconds: int = 0


class TaskListResponse(BaseModel):
    items: list[TaskListItem]


class TaskOpenRequest(BaseModel):
    source: Optional[str] = Field(default="miniapp")


class TaskOpenResponse(BaseModel):
    ok: bool
    task_id: int
    opened_at: int
    hold_seconds: int = 0

    chat_id: Optional[str] = None
    channel_post_id: Optional[int] = None
    post_url: Optional[str] = None

    session_id: Optional[str] = None


class TaskCheckRequest(BaseModel):
    session_id: Optional[str] = None


class TaskCheckResponse(BaseModel):
    ok: bool
    task_id: int
    status: Literal["completed", "already_completed", "too_early", "rejected"]
    message: str
    reward_granted: float = 0
    new_balance: float = 0
    task_completed: bool = False

File: web/lib/tasks.ts
export type TaskType = "view_post";

export type TaskStatus =
    | "available"
    | "in_progress"
    | "completed"
    | "blocked";

export type TaskCheckStatus =
    | "completed"
    | "already_completed"
    | "too_early"
    | "rejected";

export type TaskListItem = {
    id: number;
    type: TaskType;
    title: string;
    description?: string | null;
    reward: number;

    status: TaskStatus;

    chat_id?: string | null;
    channel_post_id?: number | null;
    post_url?: string | null;

    already_completed: boolean;
    can_claim: boolean;
    hold_seconds: number;
};

export type TaskOpenRequest = {
    source?: string;
};

export type TaskOpenResponse = {
    ok: boolean;
    task_id: number;
    opened_at: number;
    hold_seconds: number;
    can_check_at: number;

    chat_id?: string | null;
    channel_post_id?: number | null;
    post_url?: string | null;

    session_id?: string | null;
};

export type TaskCheckRequest = {
    session_id?: string | null;
};

export type TaskCheckResponse = {
    ok: boolean;
    task_id: number;
    status: TaskCheckStatus;
    message: string;

    reward_granted: number;
    new_balance: number;
    task_completed: boolean;
};

export type StoredOpenedTask = {
    task_id: number;
    opened_at: number;
    hold_seconds: number;
    can_check_at: number;
    session_id?: string | null;
};